text/0004-type-deduction.md

```diff
@@ -0,0 +1,214 @@
+- Feature Name: type_deduction
+- Start Date: 2020-10-13
+- RFC PR: [nushell/rfcs#0004](https://github.com/nushell/rfcs/pull/4)
+- Nushell Issue: [nushell/nushell#0000](https://github.com/nushell/nushell/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+The purpose of this RFC is to explore how type deduction can be implemented for nushell. Type deduction is helpful for e.G. aliases with typed variables, static analysis...
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently no module for the sake of type deduction is implemented. Groundwork for such a module has been layed out in alias.rs. However, the code in alias.rs doesn't handle all cases.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Given a set of variables and a block of code, the purpose of the type deduction module (TDM) is to infer all possible types a variable can have so that the block of code is still valid. In nushell a "type" is represented by a SyntaxShape variant.
+The TDM infers by looking in which kind of expression (following: containing expression) a variable is used. Then it checks which expressions (following: sub expressions) are allowed inside the containing expression at the position of the variable and maps them to their corresponding SyntaxShape variant.
+For most cases it is sufficient to look at the nushell grammar, to figure out the allowed sub-expressions. E.G. `echo 1..$var` $var is of type Int as only Int is allowed in a range expression.
+Special cases are listed below.
+
+## As a commands mandatory positional argument | As a argument to flag 
+```shell Example positional argument
+ls $path | where $filter
+$path -> SyntaxShape::FilePath 
+$filter -> SyntaxShape::Math
+```
+```shell Example named argument
+cal --full-year $year
+$year -> SyntaxShape::Int
+```
+
+The shape of a variable used as a positional argument can be infered from the command signature.
+If the signature of the command is not available, no inference will be done.
```

Comment:
```shell
external_cmd $var; ls $var
```
Infering SyntaxShape::String will clash with the later deduction when merging (empty intersection). I havent looked at the special kind of string parsing the parser does. Thanks for pointing that out. :)

Reply:
I can't remember _exactly_ what it does, but I think it basically parses everything exactly as is (in other words, quotes and everything are kept as is).

I realized I should have linked you here: https://github.com/nushell/nushell/blob/8453261211eb0822ff2253c1a682975601fffdc8/crates/nu-cli/src/commands/classified/external.rs#L64-L88

That's the code that evaluates any argument to an external. You can see there are two possible things that won't error:

1. any value that can `as_string()`, or
2. a `Table`, where we automatically splat that table across args, infallibly converting each entry to a string.

So I think no inference would work fine, but if you did want to infer something, perhaps the things that can [coerce to a String](https://github.com/nushell/nushell/blob/8453261211eb0822ff2253c1a682975601fffdc8/crates/nu-protocol/src/value.rs#L277-L288).

Reply:
Thinking about the case with:

```
external_cmd $var; ls $var
```
Reminds me that the commands themselves might be a bit too strict. To your example:

```
> ^ echo foo.txt | open $it
```

Should work in practice, but can't now because we're too strict about the shapes.

Reply:
> So I think no inference would work fine, but if you did want to infer something, perhaps the things that can [coerce to a String](https://github.com/nushell/nushell/blob/8453261211eb0822ff2253c1a682975601fffdc8/crates/nu-protocol/src/value.rs#L277-L288).

I don't think the added code complexity would outweigh the added benefit. Lets leave it out for now.